### PR TITLE
System state machine and beacon tasks

### PR DESCRIPTION
**Pull request by ArashYzd**
A first draft of the system state machine and beacon tasks are defined. The changes build but have not been tested otherwise.

### CODE REVIEW

File: .settings/org.eclipse.core.resources.prefs [hunk 1]
```diff
@@ -33,6 +33,8 @@ encoding//Debug/ex2_hal/ex2_sband_software/hardware_interface/subdir_rules.mk=UT
 encoding//Debug/ex2_hal/ex2_sband_software/hardware_interface/subdir_vars.mk=UTF-8
 encoding//Debug/ex2_hal/ex2_sband_software/subdir_rules.mk=UTF-8
 encoding//Debug/ex2_hal/ex2_sband_software/subdir_vars.mk=UTF-8
+encoding//Debug/ex2_hal/ex2_uhf_software/equipment_handler/include/subdir_rules.mk=UTF-8
+encoding//Debug/ex2_hal/ex2_uhf_software/equipment_handler/include/subdir_vars.mk=UTF-8
```
Comment (ArashYzd) [inline .settings/org.eclipse.core.resources.prefs:37]: UHF eq_handler is also added to the build after merging the new changes.

File: ex2_system/include/system_state/state_task.h [hunk 2]
```diff
@@ -20,9 +20,42 @@
...
 #include "main/system.h"
 
+enum SAT_state {
+  hw_critical_state = 0,  // EPS will shut OBC down prior to informing it
+  critical_state = 1,     // EPS will shut OBC down prior to informing it
+  safe_state = 2,
+  operational_state = 3
+};
+
+/* It might be a good idea to replace it with SAT_returnState. */
```
Comment (ArashYzd) [inline ex2_system/include/system_state/state_task.h:35]: thoughts? (Please check the .c file for the use case)

File: ex2_system/source/beacon/beacon_task.c [hunk 3]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
...
-    for ( ;; ) {
-        // TODO construct and send the system beacon
+static void *beacon_daemon(void *pvParameters) {
+  TickType_t delay = pdMS_TO_TICKS(1000);  // make this config.able
+  for (;;) {
+    int8_t uhf_status = -1;
+    /* Constructing the system beacon content */
+    // Refer to table 3 of MOP
+    UHF_configStruct beacon_msg;
+    // TODO: call the appropriate HAL functions to get the most updated or
```
Comment (ArashYzd) [inline ex2_system/source/beacon/beacon_task.c:42]: Probably sth we'll come back to in a few months. Not certain information right now.

File: ex2_system/source/beacon/beacon_task.c [hunk 4]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
...
+    // cached information of the components + state machine, RTC, etc.
```
Comment (ArashYzd) [inline ex2_system/source/beacon/beacon_task.c:43]: Probably sth we'll come back to in a few months. Not certain information right now.

File: ex2_system/source/beacon/beacon_task.c [hunk 5]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
...
+    // The beacon transmission period is configurable through comms service
+    // by the operator or here through HAL_UHF_getBeaconT().
+    uint8_t scw[SCW_LEN];
+    uhf_status = HAL_UHF_getSCW(scw);
+    if (uhf_status == U_GOOD_CONFIG) {  // replace 0 with no error status
+      scw[5] = 1;
+      uhf_status = HAL_UHF_setSCW(scw);
+    }
+    if (uhf_status != 0) {  // If get or set scw fail
+                            // Change delay? reset TRX?
```
Comment (ArashYzd) [inline ex2_system/source/beacon/beacon_task.c:57]: Do we really have any plan for when a component is turned off when it's not supposed to?

File: ex2_system/source/system_state/state_task.c [hunk 6]
```diff
@@ -50,12 +84,167 @@ static void state_daemon(void *pvParam) {
...
+      if (is_low_pwr) {
+        // send any command to wake it up, e.g.,
+        uint8_t scw[12];
+        status = HAL_UHF_getSCW(&scw);
+        // Here write any code based on the read scw!
+      }
+    }
+    if (status != 0) {
+      ex2_log("Switching UHF TRX's power failed.");
+      // An error occurred. Call this function again?
```
Comment (ArashYzd) [inline ex2_system/source/system_state/state_task.c:184]: Thoughts?

**Comment by ArashYzd** [comment]
@arrooney I've left some comments. Please let me know if you have any concerns or pieces of advice to improve the code.

p.s. Let's add this repo to #sw_ull_requests.

File: ex2_system/source/beacon/beacon_task.c [hunk 7]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
  * @param pvParameters
  *    task parameters (not used)
  */
-static void * beacon_daemon(void *pvParameters) {
-    TickType_t delay = pdMS_TO_TICKS(1000);
-    for ( ;; ) {
-        // TODO construct and send the system beacon
+static void *beacon_daemon(void *pvParameters) {
+  TickType_t delay = pdMS_TO_TICKS(1000);  // make this config.able
```
[the remark this answers is not in the document]
Comment (arrooney) [reply ex2_system/source/beacon/beacon_task.c:36]: Make delay be a static global variable that can be set from outside the beacon task

File: ex2_system/source/beacon/beacon_task.c [hunk 8]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
...
-        vTaskDelay(delay);
+    uhf_status = HAL_UHF_setBeaconMsg(beacon_msg);
+
+    /* Sending the beacon */
+    // The beacon transmission period is configurable through comms service
+    // by the operator or here through HAL_UHF_getBeaconT().
+    uint8_t scw[SCW_LEN];
+    uhf_status = HAL_UHF_getSCW(scw);
+    if (uhf_status == U_GOOD_CONFIG) {  // replace 0 with no error status
+      scw[5] = 1;
```
Comment (arrooney) [inline ex2_system/source/beacon/beacon_task.c:53]: Make these not magic numbers

File: ex2_system/source/beacon/beacon_task.c [hunk 9]
```diff
@@ -32,13 +32,33 @@ SAT_returnState start_beacon_daemon(void);
...
+      uhf_status = HAL_UHF_setSCW(scw);
+    }
+    if (uhf_status != 0) {  // If get or set scw fail
```
Comment (arrooney) [inline ex2_system/source/beacon/beacon_task.c:56]: != U_GOOD_CONFIG

File: ex2_system/source/beacon/beacon_task.c [hunk 10]
```diff
@@ -48,13 +68,11 @@ static void * beacon_daemon(void *pvParameters) {
...
 SAT_returnState start_beacon_daemon(void) {
-    if (xTaskCreate((TaskFunction_t)beacon_daemon,
-                  "coordinate_management_daemon", 2048, NULL, BEACON_TASK_PRIO,
-                  NULL) != pdPASS) {
-        ex2_log("FAILED TO CREATE TASK coordinate_management_daemon\n");
-        return SATR_ERROR;
-    }
-    ex2_log("Coordinate management started\n");
-    return SATR_OK;
+  if (xTaskCreate((TaskFunction_t)beacon_daemon, "coordinate_management_daemon",
```
Comment (arrooney) [inline ex2_system/source/beacon/beacon_task.c:71]: rename to "beacon_daemon"

File: ex2_system/source/system_state/state_task.c [hunk 11]
```diff
@@ -21,25 +21,59 @@
...
  * Query state from NanoAvionics EPS and make required updates
  * as-per the modes of operations.
  *
  * @param param
  *  Task parameter (not used)
+ * @attention OBC is not operational in HW critical and critical states
  */
 static void state_daemon(void *pvParam) {
-  TickType_t delay = pdMS_TO_TICKS(1000);
+  TickType_t state_delay = pdMS_TO_TICKS(1000);
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:45]: same here, make this a static global that can be set from outside

File: ex2_system/source/system_state/state_task.c [hunk 12]
```diff
@@ -21,25 +21,59 @@
...
+    SAT_state_e SAT_state;
+    uint8_t system_ctrl =
+        1;  // A control word deciding what systems must be working
+    // LSB -> MSB: Battery heaters, EPS, OBC, UHF, S-band, Iris, DFGM, ADCS
+    eps_batt_mode = get_eps_batt_mode();
+    SAT_state = eps2sat_mode_cnv(eps_batt_mode);
+
+    switch (SAT_state) {
+      case hw_critical_state:
+        system_ctrl = 0b00000001;  // Battery heaters on
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:58]: BATTERY_HEATERS=0b00000001
EPS_POWER=0b00000010

system_ctrl = BATTERY_HEATERS  |  EPS_POWER

File: ex2_system/source/system_state/state_task.c [hunk 13]
```diff
@@ -50,12 +84,167 @@ static void state_daemon(void *pvParam) {
  *  error report of creation
  */
 SAT_returnState start_state_daemon() {
-  if (xTaskCreate((TaskFunction_t)state_daemon,
-                "state_task", 512, NULL, STATE_TASK_PRIO,
-                NULL) != pdPASS) {
+  if (xTaskCreate((TaskFunction_t)state_daemon, "state_task", 512, NULL,
+                  STATE_TASK_PRIO, NULL) != pdPASS) {
     ex2_log("FAILED TO CREATE TASK state_task\n");
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:89]: change to printf

File: ex2_system/source/system_state/state_task.c [hunk 14]
```diff
@@ -50,12 +84,167 @@ static void state_daemon(void *pvParam) {
...
+SAT_state_e eps2sat_mode_cnv(eps_mode_e batt_mode) {
+  SAT_state_e state;
+  switch (batt_mode) {
+    case critical:
+      state = critical_state;
+      break;
+    case safe:
+      state = safe_state;
+      break;
+    case normal:
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:113]: just copy the code below

File: ex2_system/source/system_state/state_task.c [hunk 15]
```diff
@@ -50,12 +84,167 @@ static void state_daemon(void *pvParam) {
...
+ * @brief Switches the power of each subsystem.
+ * @params ctrl_word
+ *  system control word generated by state machine
+ * @attention The operations functions must be accessible to call from a GS
+ * command too.
+ * @attention May be replaced with an object
+ * @returns
+ */
+void change_systems_status(uint8_t ctrl_word) {
+  int8_t sys_state[8] = {1};
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:135]: define 8 to be something (number of switchable subsystems?)

File: ex2_system/source/system_state/state_task.c [hunk 16]
```diff
@@ -50,12 +84,167 @@ static void state_daemon(void *pvParam) {
...
+  systems_status systems_ctrl;  //* Can be made global to be accessed by HK
+  systems_ctrl.EPS = (ctrl_word >> 1) & 1;
+  systems_ctrl.OBC = (ctrl_word >> 2) & 1;
+  systems_ctrl.UHF = (ctrl_word >> 3) & 1;
+  sys_state[3] = power_switch_uhf(&systems_ctrl.UHF);
```
Comment (arrooney) [inline ex2_system/source/system_state/state_task.c:140]: define each of these index values to indicate the subsystem they represent

File: ex2_system/include/system_state/state_task.h (hunk 2 above)
Comment (arrooney) [reply ex2_system/include/system_state/state_task.h:35]: I think this is fine!

**Comment by ArashYzd** [comment]
@arrooney Let me know if it's good to go. I'll push a tiny commit to eps repo, too.

**Pull request merged**